### PR TITLE
Task genrator debug api

### PR DESCRIPTION
**Pull request by saurabhd336**
Save task generator info into an inmemory map, and provide a controller API to fetch the details for a given table
design doc: https://docs.google.com/document/d/1bCn13k57WLSaIAQGfATP75VS3p6V_ievlBi_HoOtmZ8/edit?usp=sharing

### CODE REVIEW

**Comment by saurabhd336** [comment]
@npawar @mcvsubbu I've raised this PR in place of the now closed PR https://github.com/apache/pinot/pull/9043

**Comment by mcvsubbu** [comment]
Overall, I am a plus 1 on more minion debuggability. 

My view is that we should add APIs to minions similar to that of servers, and pull debugging info from there and display it.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java [hunk 1]
```diff
@@ -517,7 +526,34 @@ private synchronized Map<String, String> scheduleTasks(List<String> tableNamesWi
...
+              taskGeneratorMostRecentRunInfo -> taskGeneratorMostRecentRunInfo.addSuccessRunTs(
+                  System.currentTimeMillis()));
+        } catch (Exception exception) {
+          LOGGER.warn("Failed to save task generator success timestamp to ZK", exception);
+        }
+      }
+    } catch (Exception e) {
+      for (TableConfig tableConfig : enabledTableConfigs) {
+        try {
+          StringWriter errors = new StringWriter();
```
Comment (npawar) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:544]: it seems we will do the same thing for each table config, and then put the same stacktrace for every table?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/helix/FakePropertyStore.java [hunk 2]
```diff
@@ -37,7 +37,13 @@ public FakePropertyStore() {
 
   @Override
   public ZNRecord get(String path, Stat stat, int options) {
-    return _contents.get(path);
+    ZNRecord znRecord = _contents.get(path);
```
Comment (npawar) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/helix/FakePropertyStore.java:40]: is this change needed?

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      // sort first in case the given timestamp is not the largest one.
+      Collections.sort(_mostRecentSuccessRunTS);
+      _mostRecentSuccessRunTS.remove(0);
+    }
+  }
+
+  /**
+   * Returns the current information version, it should be consistent with the corresponding {@link ZNRecord} version.
+   */
+  public int getVersion() {
```
Comment (npawar) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:137]: prolly don't need this method either

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  @Nonnull
+  private final List<Long> _mostRecentSuccessRunTS;
+  private final int _version;
+
+  private TaskGeneratorMostRecentRunInfo(String tableNameWithType, String taskType,
+      Map<Long, String> mostRecentErrorRunMessage, List<Long> mostRecentSuccessRunTS, int version) {
+    _tableNameWithType = tableNameWithType;
+    _taskType = taskType;
+    // sort and keep the most recent several error messages
+    _mostRecentErrorRunMessage = new TreeMap<>();
```
Comment (npawar) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:61]: it seems this constructor is only ever called from within this same class, with empty list and map. Do we need all this sorting logic?

Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:61]: If at all we move to persistent storage, we might need to serialize / deserialize instances of this class using an allArgsConstructor

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java (hunk 1 above)
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:544]: Fixed.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/InMemoryTaskManagerStatusCache.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
+package org.apache.pinot.common.minion;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (klsince) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/InMemoryTaskManagerStatusCache.java:9]: format: move this header to top

**Comment by mcvsubbu** [comment]
> > Overall, I am a plus 1 on more minion debuggability.
> > My view is that we should add APIs to minions similar to that of servers, and pull debugging info from there and display it.
> 
> We already have minion APIs (look for /debug endpoint in Tasks tab). This is for scheduler errors, and the scope of scheduler remains limited to the controller, hence thought it makes most sense as an API here

I don't see a minion debug REST endpoint in the code. Am I missing something?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java [hunk 6]
```diff
@@ -231,6 +236,23 @@ public Map<String, PinotHelixTaskResourceManager.TaskDebugInfo> getTasksDebugInf
...
 
+  @GET
+  @Path("/tasks/generator/{tableNameWithType}/{taskType}/debug")
+  @ApiOperation("Fetch task generation information for the recent runs of the given task for the given table")
+  public BaseTaskGeneratorInfo getTaskGenerationDebugInto(
+      @ApiParam(value = "Task type", required = true) @PathParam("taskType") String taskType,
+      @ApiParam(value = "Table name with type", required = true) @PathParam("tableNameWithType")
+          String tableNameWithType) {
+    BaseTaskGeneratorInfo
+        taskGeneratorMostRecentRunInfo = _taskManagerStatusCache.fetchTaskGeneratorInfo(tableNameWithType, taskType);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: What if the table is not mastered on this controller?

**Review by mcvsubbu — COMMENTED** [review]
@saurabhd336  thanks a lot for your contribution. It has been hard to debug minions, so I totally understand your pain point.  However, I have a few general questions:
1. What is the use case where you think this will help? (we have some use cases where we saw problems, and instead of introducing one new API per use case, I am looking to make sure we do something that works for more than one. @sajjad-moradi  can review from our standpoint).
2. I know @npawar referred to a debug API on the minion, but I could not locate one. Please correct me if I am wrong. I think we should issue an API to the minions as well (liek we do with servers) and get the latest status form them and display it in the controller.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java (hunk 6 above)
Comment (npawar) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: > @saurabhd336 thanks a lot for your contribution. It has been hard to debug minions, so I totally understand your pain point. However, I have a few general questions:
> 
> 1. What is the use case where you think this will help? (we have some use cases where we saw problems, and instead of introducing one new API per use case, I am looking to make sure we do something that works for more than one. @sajjad-moradi  can review from our standpoint).
> 2. I know @npawar referred to a debug API on the minion, but I could not locate one. Please correct me if I am wrong. I think we should issue an API to the minions as well (liek we do with servers) and get the latest status form them and display it in the controller.

Hey Subbu, this is the one for Minion tasks: `/tasks/task/{taskName}/debug`. In the list of subtasks returned, there's an "info" field which contains the exception (if any) encountered by this task

Comment (npawar) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: ah good catch about the lead controller! this won't work as is, we'll have to think of how to handle that

Comment (klsince) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: for 2. I feel it's not very convenient when one auto scales minion workers frequently based on pending tasks. The debug API would be either not reachable when no workers are around or get no past task states to check when new workers are added

Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: But we have helix APIs for that, right? Has that avenue been fully explored? We are migrating to Helix 1.x) very soon (we are getting to production with that), so please check in 1.x pages what is available for task management and debugging, thanks. We will also do the same.
cc: @sajjad-moradi

Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: There are two parts:
- generating task configurations which happens in pinot code
- submitting the task via TaskDriver which is in Helix code
Subbu has a valid point that helix should provide some info on failure on the 2nd part - task submission. The first part though is before reaching helix and it's good if we have an API to capture issues in task generation in pinot code.

Comment (sajjad-moradi) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:247]: Also scheduling tasks can be done by either:
- periodic task generator job on lead controller, or
- task REST endpoint which can be on any controller

So this means the approach in this PR doesn't work as expected.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/InMemoryTaskManagerStatusCache.java (hunk 5 above)
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/InMemoryTaskManagerStatusCache.java:2]: Ack

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  public List<Long> getMostRecentSuccessRunTS() {
+    return _mostRecentSuccessRunTS;
+  }
+
+  /**
+   * Adds a success task generating run timestamp
+   * @param ts A timestamp
+   */
+  public void addSuccessRunTs(long ts) {
+    _mostRecentSuccessRunTS.add(ts);
```
Comment (klsince) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:126]: Q: any need to keep _mostRecentSuccessRunTS sorted?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java [hunk 8]
```diff
@@ -517,7 +526,34 @@ private synchronized Map<String, String> scheduleTasks(List<String> tableNamesWi
   private String scheduleTask(PinotTaskGenerator taskGenerator, List<TableConfig> enabledTableConfigs,
       boolean isLeader) {
     LOGGER.info("Trying to schedule task type: {}, isLeader: {}", taskGenerator.getTaskType(), isLeader);
-    List<PinotTaskConfig> pinotTaskConfigs = taskGenerator.generateTasks(enabledTableConfigs);
+    List<PinotTaskConfig> pinotTaskConfigs;
+    try {
+      pinotTaskConfigs = taskGenerator.generateTasks(enabledTableConfigs);
+      for (TableConfig tableConfig : enabledTableConfigs) {
+        try {
+          _taskManagerStatusCache.saveTaskGeneratorInfo(tableConfig.getTableName(), taskGenerator.getTaskType(),
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:534]: nit: add a saveTaskGeneratorInfoQuietly() to save the try-catch block for brevity.

**Review by klsince — APPROVED** [review]
LGTM

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java (hunk 4 above)
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:61]: Since we decided to not use ZNode to store the information, suggest simplifying it to only take `tableNameWithType` and `taskType`. We may change it when we decide to persist the info in the future.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  @VisibleForTesting
+  static final String MOST_RECENT_SUCCESS_RUN_TS = "mostRecentSuccessRunTs";
+  @VisibleForTesting
+  static final String MOST_RECENT_ERROR_RUN_MESSAGE = "mostRecentErrorRunMessage";
+
+  private final String _taskType;
+  private final String _tableNameWithType;
+  // the timestamp to error message map of the most recent several error runs
+  @Nonnull
+  private final TreeMap<Long, String> _mostRecentErrorRunMessage;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:50]: Consider storing timestamp as `java.sql.Timestamp` so that the response can show human readable time

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java [hunk 10]
```diff
@@ -231,6 +260,55 @@ public Map<String, PinotHelixTaskResourceManager.TaskDebugInfo> getTasksDebugInf
...
+        new CompletionServiceHelper(_executor, _connectionManager, HashBiMap.create(0));
+    CompletionServiceHelper.CompletionServiceResponse serviceResponse =
+        completionServiceHelper.doMultiGetRequest(controllerUrls, null, true, 10000);
+
+    List<JsonNode> result = new ArrayList<>();
+    serviceResponse._httpResponses.values().forEach(resp -> {
+      try {
+        result.add(JsonUtils.stringToJsonNode(resp));
+      } catch (IOException e) {
+        LOGGER.error("Failed to parse ");
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:305]: Incomplete

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java [hunk 11]
```diff
@@ -517,7 +526,34 @@ private synchronized Map<String, String> scheduleTasks(List<String> tableNamesWi
   private String scheduleTask(PinotTaskGenerator taskGenerator, List<TableConfig> enabledTableConfigs,
       boolean isLeader) {
     LOGGER.info("Trying to schedule task type: {}, isLeader: {}", taskGenerator.getTaskType(), isLeader);
-    List<PinotTaskConfig> pinotTaskConfigs = taskGenerator.generateTasks(enabledTableConfigs);
+    List<PinotTaskConfig> pinotTaskConfigs;
+    try {
+      pinotTaskConfigs = taskGenerator.generateTasks(enabledTableConfigs);
+      for (TableConfig tableConfig : enabledTableConfigs) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:532]: Let's add a `TODO` here to separate the generated task/exception for each table. Currently even though some table doesn't have anything scheduled (e.g. enough tasks are scheduled from previous tables), we will still put success TS for it

File: pinot-common/src/main/java/org/apache/pinot/common/minion/InMemoryTaskManagerStatusCache.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+package org.apache.pinot.common.minion;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+public class InMemoryTaskManagerStatusCache extends TaskManagerStatusCache<TaskGeneratorMostRecentRunInfo> {
+
+  private static class TaskGeneratorCacheKey {
+    String _tableNameWithType;
```
[the remark this answers is not in the document]
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/InMemoryTaskManagerStatusCache.java:29]: Ack

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+import java.util.Queue;
+import java.util.TreeMap;
+import javax.annotation.Nonnull;
+import org.apache.helix.ZNRecord;
+
+
+/**
+ * a task generator running history which keeps the most recent several success run timestamp and the most recent
+ * several error run
+ * message.
```
[the remark this answers is not in the document]
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:36]: Ack

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+ */
+public class TaskGeneratorMostRecentRunInfo extends BaseTaskGeneratorInfo {
+  @VisibleForTesting
+  static final int MAX_NUM_OF_HISTORY_TO_KEEP = 5;
+  @VisibleForTesting
+  static final String MOST_RECENT_SUCCESS_RUN_TS = "mostRecentSuccessRunTs";
+  @VisibleForTesting
+  static final String MOST_RECENT_ERROR_RUN_MESSAGE = "mostRecentErrorRunMessage";
```
[the remark this answers is not in the document]
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:44]: This was being used when we had ZK based storage. Removed

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+  private final String _taskType;
+  private final String _tableNameWithType;
+  // the timestamp to error message map of the most recent several error runs
+  @Nonnull
```
[the remark this answers is not in the document]
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:49]: Ack

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java (hunk 4 above)
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:61]: Ack

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java (hunk 7 above)
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:126]: Simplifies the eviction is all. Also the APi response is easier to work with (the most recent timestamp is the one of the top)

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java (hunk 10 above)
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:305]: Ack

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java (hunk 8 above)
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:534]: With the in-memory storage, I don't see any exceptions that can be thrown. Have removed the try catch block

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java (hunk 11 above)
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:532]: Didn't get this. This method is only called for `enabledTableConfigs` i.e. only for tables for which the particular task is enabled?

Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:537]: Had a discussion offline. Added the todo

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+    return _taskType;
+  }
+
+  /**
+   * Gets the timestamp to error message map of the most recent several error runs
+   */
+  public TreeMap<String, String> getMostRecentErrorRunMessages() {
+    TreeMap<String, String> result = new TreeMap<>();
+    _mostRecentErrorRunMessages.forEach((timestamp, error) -> result.put(
+        OffsetDateTime.ofInstant(Instant.ofEpochMilli(timestamp), ZoneId.of("UTC")).toString(), error));
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:73]: (minor)
```suggestion
        OffsetDateTime.ofInstant(Instant.ofEpochMilli(timestamp), ZoneOffset.UTC).toString(), error));
```

We can also consider putting `ZoneId.systemDefault()` which might be more readable, but it might also be confusing when the client is in a different timezone from the server. IMO both ways have pros and cons, so your call

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskManagerStatusCache.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.minion;
+
+import java.util.function.Consumer;
+
+
+public abstract class TaskManagerStatusCache<T extends BaseTaskGeneratorInfo> {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/TaskManagerStatusCache.java:24]: Suggest changing it to an interface

**Review by Jackie-Jiang — APPROVED** [review]
LGTM

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java (hunk 16 above)
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:73]: I keeping everything UTC might be more user friendly especially when pinot clusters are deployed in different geographies. Keeping it UTC for now.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskManagerStatusCache.java (hunk 17 above)
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskManagerStatusCache.java]: Ack

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java [hunk 18]
```diff
@@ -231,6 +260,55 @@ public Map<String, PinotHelixTaskResourceManager.TaskDebugInfo> getTasksDebugInf
...
+      if (taskGeneratorMostRecentRunInfo == null) {
+        throw new ControllerApplicationException(LOGGER, "Task generation information not found",
+            Response.Status.NOT_FOUND);
+      }
+
+      return JsonUtils.objectToString(taskGeneratorMostRecentRunInfo);
+    }
+
+    // Call all controllers
+    List<InstanceConfig> controllers = _pinotHelixResourceManager.getAllControllerInstanceConfigs();
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:286]: you don't need to send request to all controllers. Just to the lead controller is enough

**Comment by mcvsubbu** [comment]
I had asked some questions about the use case you are tackling. Can you please answer those? How frequently are you generating tasks (in your use case)? Is a metric bump not enough for your use case? Can log processing help solving this use case?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java (hunk 18 above)
Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:286]: Since POST '/tasks/schedule' API lets one schedule tasks for a given table on any controller (non leader included), we do need to query all controllers here.

**Comment by saurabhd336** [comment]
@mcvsubbu 
1) @npawar @Jackie-Jiang ? I might have just very rough, possibly inaccurate numbers.

2) I feel the need of a control plane level API within pinot to give an overall view into current and past state of minion tasks is of importance to us. Task generator being a key part of the entire minion task flow. While metrics can help to some extent, having details like failure stack traces etc might be difficult. This api avoids having to tally metrics and debug logs from a separate log processing system. 

3) I suppose it could. But integrating the log processing framework into pinot APIs themselves might be a bit of a challenge. Having a system table for these kind of usecases might be the right way forward, such that pinot itself can store and serve debug data and status metrics for each of the components / flows. Essentially move from in-memory storage of logs and metrics into the system table

@npawar @Jackie-Jiang to add more

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java (hunk 18 above)
Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:286]: Perhaps I was not clear. You can field the request in any controller (like you do). If that is not the lead for the table, then locate the leader controller (see the class `PinotLeadControllerRestletResource` for how to do that) and then forward the request to just the lead controller.

**Comment by mcvsubbu** [comment]
> @mcvsubbu
> 
> 1. @npawar @Jackie-Jiang ? I might have just very rough, possibly inaccurate numbers.
> 2. I feel the need of a control plane level API within pinot to give an overall view into current and past state of minion tasks is of importance to us. Task generator being a key part of the entire minion task flow. While metrics can help to some extent, having details like failure stack traces etc might be difficult. This api avoids having to tally metrics and debug logs from a separate log processing system.

I agree there is a need to debug minions. However this API does not provide any new insights other than what Helix already provides. So, let us work off of some concrete examples -- what is the use case in which you think this will help?

> 3. I suppose it could. But integrating the log processing framework into pinot APIs themselves might be a bit of a challenge. Having a system table for these kind of usecases might be the right way forward, such that pinot itself can store and serve debug data and status metrics for each of the components / flows. Essentially move from in-memory storage of logs and metrics into the system table

Every installation has its own way of processing logs. Cloud installations typically pipe the logs into a log processing service to get insights into it. On-prem installations have either their own way of piping the logs, or ways to get onto the host and look for the logs. 

> 
> @npawar @Jackie-Jiang to add more

The reason I am reluctant to add more APIs is that the more we add, the more we need to keep backward compatibility on. My question to @saurabhd336  is : Have you looked at what Helix has to offer? If so, what are the short-comings?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java (hunk 18 above)
Comment (mcvsubbu) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:286]: Ah ok, so the POST actually has _any_ controller schedule the job for a table ? Hmm... maybe we should fix hat too. But I see what you are saying here. My bad.

**Comment by npawar** [comment]
> @mcvsubbu
> 
> 1. @npawar @Jackie-Jiang ? I might have just very rough, possibly inaccurate numbers.
> 2. I feel the need of a control plane level API within pinot to give an overall view into current and past state of minion tasks is of importance to us. Task generator being a key part of the entire minion task flow. While metrics can help to some extent, having details like failure stack traces etc might be difficult. This api avoids having to tally metrics and debug logs from a separate log processing system.
> 3. I suppose it could. But integrating the log processing framework into pinot APIs themselves might be a bit of a challenge. Having a system table for these kind of usecases might be the right way forward, such that pinot itself can store and serve debug data and status metrics for each of the components / flows. Essentially move from in-memory storage of logs and metrics into the system table
> 
> @npawar @Jackie-Jiang to add more

for 1: typically for users of RealtimeToOfflineTask, tasks get generated hourly. For SegmentGenerationAndPushTask, it can be way more frequently, depending on the the number of times files are generated in the source dir. MergeRollupTasks, are less frequent, but still several a day. There might be others, but this is what we see most commonly setup by users in oss. In a typical setup, all of these would be configured. It becomes quite confusing for users to have to find the exact exception in the logs, especially because some logs are in controller (scheduler related) and some in minion (task execution related). This API will help us make the feedback loop quicker, especially when we add this into the new Minion tab on the Pinot Admin UI

Regarding info already in Helix, these scheduler related exceptions are not present in the Helix generated metadata.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/BaseTaskGeneratorInfo.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+package org.apache.pinot.common.minion;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * Base abstract class for task generator info.
+ */
+public abstract class BaseTaskGeneratorInfo {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/BaseTaskGeneratorInfo.java:28]: Do you want to add `implements Serializable`

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java (hunk 7 above)
Comment (mcvsubbu) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:105]: Does this indicate a successful run (completion) of the task, or a successful generation of the task? If the task fails in minion, does your API address it?

**Comment by mcvsubbu** [comment]
> > @mcvsubbu
> > 
> > 1. @npawar @Jackie-Jiang ? I might have just very rough, possibly inaccurate numbers.
> > 2. I feel the need of a control plane level API within pinot to give an overall view into current and past state of minion tasks is of importance to us. Task generator being a key part of the entire minion task flow. While metrics can help to some extent, having details like failure stack traces etc might be difficult. This api avoids having to tally metrics and debug logs from a separate log processing system.
> > 3. I suppose it could. But integrating the log processing framework into pinot APIs themselves might be a bit of a challenge. Having a system table for these kind of usecases might be the right way forward, such that pinot itself can store and serve debug data and status metrics for each of the components / flows. Essentially move from in-memory storage of logs and metrics into the system table
> > 
> > @npawar @Jackie-Jiang to add more
> 
> for 1: typically for users of RealtimeToOfflineTask, tasks get generated hourly. For SegmentGenerationAndPushTask, it can be way more frequently, depending on the the number of times files are generated in the source dir. MergeRollupTasks, are less frequent, but still several a day. There might be others, but this is what we see most commonly setup by users in oss. In a typical setup, all of these would be configured. It becomes quite confusing for users to have to find the exact exception in the logs, especially because some logs are in controller (scheduler related) and some in minion (task execution related). This API will help us make the feedback loop quicker, especially when we add this into the new Minion tab on the Pinot Admin UI
> 
> Regarding info already in Helix, these scheduler related exceptions are not present in the Helix generated metadata.

I still think log processing should be the answer to this (and other similar PRs that may come up in the future). We should not be adding a new API for every error condition we may encounter in the system (and log something).

@siddharthteotia , @npawar , @Jackie-Jiang , @snleee, @kishoreg, @mayankshriv  what do you think? If the PMCs don't have any objection to this then I can live with this, but I am willing to bet that more such PRs will come up because logs are difficult to read.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java (hunk 7 above)
Comment (saurabhd336) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/TaskGeneratorMostRecentRunInfo.java:105]: This only indicates the successful generation of a the task. This API is exclusively for task generation
